### PR TITLE
chore(yamux): improve performance with zero allocation queue

### PR DESCRIPTION
**Pull request by vladopajic**
this pr improves yamux performance by 20-40% as observed when running perf test. 

previously data was constantly appended to `recvQueue` which caused memory trashing due to resizing. 
change from this pr introduced `ZeroQueue` that will allow data to be appended and consumed with zero allocations.

### CODE REVIEW

File: libp2p/utils/zeroqueue.nim [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+proc pop*(q: var ZeroQueue, count: int): seq[byte] =
+  if q.data.len == 0 or count == 0:
+    return @[]
+
+  let first = q.data[0]
+
+  # first frame has up to requested count elements,
+  # queue will return this frame (frame might have less then requested)
+  if first.len <= count:
+    q.data = q.data[1 ..^ 1]
```
Comment (arnetheduck) [inline libp2p/utils/zeroqueue.nim]: fyi, this will copy all the little queues in the big queue - `[]` allocates a new `seq`, then copies the range one by one and so on - ditto all the other places where it gets used, ie `push` will copy `b`.

In general, it just so happens that https://github.com/status-im/nim-faststreams/pull/71 was pushed just now, exposing the queue from faststreams in a semi-consumable way - it does something similar that you want to do here, but with fewer allocations and copies all around. for example the trick to not make massive copies is to reuse seqs and use a `deque` instead of a seq to manage the forward motion.

Comment (vladopajic) [reply libp2p/utils/zeroqueue.nim]: > fyi, this will copy all the little queues in the big queue - [] allocates a new seq, then copies the range one by one and so on - ditto all the other places where it gets used, ie push will copy b.

if i understand it correctly there is some trashing when `q.data` is changed with push and pop. but 2nd dimension elements are reused, right?

Comment (arnetheduck) [reply libp2p/utils/zeroqueue.nim]: no reuse whatsoever. `[]` always creates a full deep copy of everything but the zero-eth element that you're discarding. if you had 1000 seqs in q.data, and each had 1000 bytes, this will allocate 999000 bytes (plus rounding) and then throw them away after the assignment.

With a `seq`, you can never remove anything from the front without _at least_ moving data around - it's the equivalent of a C++ `vector`. Sometimes you can avoid copying the "inner" elements, but the way you've written it, they are really allocated anew.

[`q.data.delete(0)`](https://nim-lang.org/docs/system.html#delete,seq[T],Natural) would move the elements in the seq. ie in the above example, we would not allocate any additional data, but we'd still have to perform 999 moves, no copies and no allocations. With a [`Deque[seq[byte]]`](https://nim-lang.org/docs/deques.html) you don't perform any moves or copies apart from that first item you're popping out of there.

Comment (arnetheduck) [reply libp2p/utils/zeroqueue.nim]: fwiw, `PageBuffers` in faststreams were created for the same problem that you're working on here. I'd recommend taking a look and maybe trying to improve PB so we can reuse the same general buffer for multiple projects. I think libp2p already depends on faststreams - or it will when using `nim-protobuf-serialization` which uses faststreams under the hood.

Comment (arnetheduck) [reply libp2p/utils/zeroqueue.nim]: (and as a second note, we've outlawed the use of `[a..b]` in everything but tests - if you want a read-only view, `q.data.toOpenArray(a, b)` is the way but that's not what you want to use here (since using it would imply a copy instead of a move)

File: libp2p/utils/zeroqueue.nim [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+proc len*(q: ZeroQueue): int64 =
+  var l: int64
+  for b in q.chunks:
+    l += b.len()
+  return l
+
+proc push*(q: var ZeroQueue, b: seq[byte]) =
+  if b.len > 0:
+    q.chunks.add(Chunk(address: addr b[0], size: b.len, start: 0))
```
Comment (arnetheduck) [inline libp2p/utils/zeroqueue.nim]: this idea unfortunately will lead to a crash or invalid data. if `b` is garbage-collected before the chunk is removed, the program will crash (or write some random data). 

Generally, `ptr` is used when the lifetime of the pointee is managed by "something", such as a `ref`, and that something outlives the `ptr` itself.

Comment (vladopajic) [reply libp2p/utils/zeroqueue.nim]: yes, could be fixed by adding `ref` to the `Chunk`  👽

File: libp2p/utils/zeroqueue.nim (hunk 1 above)
Comment (vladopajic) [reply libp2p/utils/zeroqueue.nim]: got it

i'll have a look at `PageBuffers` tomorrow, i overdid my screen time for today.

on the side note, i  don't really like this approach although it does improve performance here. each layer will introduces performance loss where data has to be copied. it would be much better if ownership of data (starting from origin) is transferred up the layers.  iam trying to say it still doesn't feel like it's right way to do..

Comment (arnetheduck) [reply libp2p/utils/zeroqueue.nim]: indeed, this has long been an issue in nim - there is some work done to improve the situation where the "last use" of a seq automatically makes it eligible for moving instead of copying the data - this would be useful in a buffer like this, though it needs an additional `sink` annotation - see https://nim-lang.org/docs/destructors.html#sink-parameters

File: libp2p/muxers/yamux/yamux.nim [hunk 3]
```diff
@@ -252,12 +252,12 @@ proc updateRecvWindow(
   ##
   # In order to avoid spamming a window update everytime a byte is read,
   # we send it everytime half of the maxRecvWindow is read.
-  let inWindow = channel.recvWindow + channel.recvQueue.len
+  let inWindow = channel.recvWindow + channel.recvQueue.len()
```
Comment (gmelodie) [inline libp2p/muxers/yamux/yamux.nim]: nit: maybe we should adopt the same style as other teams here?
verbs: put `()`
nouns (like `len`): no parenthesis
https://github.com/vacp2p/nim-libp2p/pull/1472#discussion_r2162507398

File: libp2p/muxers/yamux/yamux.nim [hunk 4]
```diff
@@ -290,31 +290,26 @@ method readOnce*(
...
 proc gotDataFromRemote(
     channel: YamuxChannel, b: seq[byte]
 ) {.async: (raises: [CancelledError, LPStreamError]).} =
   channel.recvWindow -= b.len
-  channel.recvQueue = channel.recvQueue.concat(b)
+  channel.recvQueue.push(b)
   channel.receivedData.fire()
   when defined(libp2p_yamux_metrics):
-    libp2p_yamux_recv_queue.observe(channel.recvQueue.len.int64)
+    libp2p_yamux_recv_queue.observe(channel.recvQueue.len().int64)
```
Comment (gmelodie) [inline libp2p/muxers/yamux/yamux.nim]: as above

File: libp2p/utils/zeroqueue.nim [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+proc len*(q: ZeroQueue): int64 =
+  var l: int64
+  for b in q.chunks.items():
+    l += b.len()
+  return l
+
+proc push*(q: var ZeroQueue, b: sink seq[byte]) =
+  if b.len > 0:
+    q.chunks.addLast(newChunk(b))
```
Comment (gmelodie) [inline libp2p/utils/zeroqueue.nim:46]: What if `b.len == 0`?

File: tests/testzeroqueue.nim [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+# those terms.
+
+import unittest2
+import ../libp2p/utils/zeroqueue
+
+proc toSeq(p: pointer, length: int): seq[byte] =
+  let b = cast[ptr UncheckedArray[byte]](p)
+  result = newSeq[byte](length)
+  for i in 0 ..< length:
+    result[i] = b[i]
```
Comment (gmelodie) [inline tests/testzeroqueue.nim]: Maybe this instead? You have to guarantee that `p` has valid data in it though
```suggestion
result = newSeq[byte](length)
copyMem(result[0].addr, p, length)
```

File: tests/testzeroqueue.nim [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+suite "ZeroQueue":
+  test "push-pop":
+    var q: ZeroQueue
+    check q.len() == 0
+    check q.isEmpty()
```
Comment (gmelodie) [inline tests/testzeroqueue.nim:25]: nit
```suggestion
    check q.len() == 0 and q.isEmpty()
```

File: tests/testzeroqueue.nim [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    check q.popChunkSeq(1).len == 0 # pop empty seq when queue is empty
+
+    q.push(@[1'u8, 2, 3])
+    q.push(@[4'u8, 5])
+    check q.len() == 5
+    check not q.isEmpty()
```
Comment (gmelodie) [inline tests/testzeroqueue.nim:31]: nit
```suggestion
    check  not q.isEmpty() and q.len() == 5
```

File: tests/testzeroqueue.nim [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    defer:
+      dealloc(pbytes)
+
+    # consumeTo should fill up to nbytes (queue is emptied)
+    q.push(@[1'u8, 2, 3])
+    q.push(@[4'u8, 5])
+    q.push(@[6'u8, 7])
+    check q.consumeTo(pbytes, nbytes) == 7
+    check toSeq(pbytes, 7) == @[1'u8, 2, 3, 4, 5, 6, 7]
+    check q.isEmpty()
```
Comment (gmelodie) [inline tests/testzeroqueue.nim]: this could be a separate test

File: tests/testzeroqueue.nim [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+    # consumeTo should fill only 1 element, leaving 2 elements in the queue
+    q.push(@[1'u8, 2, 3])
+    check q.consumeTo(pbytes, 1) == 1
+    check toSeq(pbytes, 1) == @[1'u8]
+    check q.len() == 2
+    # assert remaning elements
+    check q.consumeTo(pbytes, nbytes) == 2
+    check toSeq(pbytes, 2) == @[2'u8, 3]
+    check q.isEmpty()
```
Comment (gmelodie) [inline tests/testzeroqueue.nim]: this could be a separate test

File: tests/testzeroqueue.nim [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    q.clear()
+    q.push(@[4'u8, 5])
+    q.push(@[1'u8, 2, 3])
+    check q.consumeTo(pbytes, 3) == 3
+    check toSeq(pbytes, 3) == @[4'u8, 5, 1]
+    check q.len() == 2
+    # assert remaning elements
+    check q.consumeTo(pbytes, nbytes) == 2
+    check toSeq(pbytes, 2) == @[2'u8, 3]
+    check q.isEmpty()
```
Comment (gmelodie) [inline tests/testzeroqueue.nim]: this could be a separate test

**Review by gmelodie — APPROVED** [review]

File: tests/testzeroqueue.nim (hunk 7 above)
Comment (vladopajic) [reply tests/testzeroqueue.nim:25]: not in favor of this, if check fails you will not know which statements where false

File: tests/testzeroqueue.nim (hunk 8 above)
Comment (vladopajic) [reply tests/testzeroqueue.nim:31]: not in favor of this, if check fails you will not know which statements where false

File: libp2p/utils/zeroqueue.nim (hunk 5 above)
Comment (vladopajic) [reply libp2p/utils/zeroqueue.nim:46]: there is no need in appending sequence with zero size

File: tests/testzeroqueue.nim (hunk 9 above)
Comment (vladopajic) [reply tests/testzeroqueue.nim]: could be, but there is no need. this tests tests everything `consumeTo` related

File: tests/testzeroqueue.nim (hunk 11 above)
Comment (vladopajic) [reply tests/testzeroqueue.nim]: could be, but there is no need. this tests tests everything `consumeTo` related

File: tests/testzeroqueue.nim (hunk 10 above)
Comment (vladopajic) [reply tests/testzeroqueue.nim]: could be, but there is no need. this tests tests everything `consumeTo` related

File: libp2p/utils/zeroqueue.nim [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+  c.size - c.start
+
+type ZeroQueue* = object
+  # ZeroQueue is queue structure optimized for efficient pushing and popping of 
+  # byte sequences `seq[byte]` (called chunks). This type is useful for streaming or buffering 
+  # scenarios where chunks of binary data are accumulated and consumed incrementally.
+  chunks: Deque[Chunk]
+
+proc clear*(q: var ZeroQueue) =
+  q.chunks.clear()
```
Comment (richard-ramos) [inline libp2p/utils/zeroqueue.nim:33]: Maybe to keep consistency with other procs?
```suggestion
  return q.chunks.clear()
```

File: libp2p/utils/zeroqueue.nim [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+
+proc isEmpty*(q: ZeroQueue): bool =
```
Comment (richard-ramos) [inline libp2p/utils/zeroqueue.nim:35]: Can probably be a template so this code gets inlined

File: libp2p/utils/zeroqueue.nim [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+  ret.size = ret.start + count
+  first.start += count
+  q.chunks.addFirst(first)
+  return ret
+
+proc consumeTo*(q: var ZeroQueue, pbytes: pointer, nbytes: int): int =
+  var consumed = 0
+  while consumed < nbytes and not q.isEmpty():
+    let chunk = q.popChunk(nbytes - consumed)
+    let dest = cast[pointer](cast[ByteAddress](pbytes) + consumed)
```
Comment (richard-ramos) [inline libp2p/utils/zeroqueue.nim:68]: Byteaddress are deprecated, gotta use `uint` instead:
https://nim-lang.github.io/Nim/ctypes.html#ByteAddress

File: tests/testzeroqueue.nim [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+
+# Nim-Libp2p
+# Copyright (c) 2025 Status Research & Development GmbH
+# Licensed under either of
+#  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+#  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+# at your option.
+# This file may not be copied, modified, or distributed except according to
+# those terms.
+
```
Comment (richard-ramos) [inline tests/testzeroqueue.nim:11]: Edge cases:
  1. A large chunk is partially consumed multiple times
  2. Consuming from a queue with multiple partially consumed chunks (i.e. push, partially consume, push, partially consume) to make sure that doing a consume, next push is not messed up
  3. Empty pushes are being tested in isolation, what about pushing something, partial consume, empty push, and consume.
  4. Behavior of consume on an empty queue

File: libp2p/utils/zeroqueue.nim [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+  var l: int64
+  for b in q.chunks.items():
+    l += b.len()
+  return l
+
+proc push*(q: var ZeroQueue, b: sink seq[byte]) =
+  if b.len > 0:
+    q.chunks.addLast(newChunk(b))
+
+proc popChunk(q: var ZeroQueue, count: int): Chunk {.inline.} =
```
Comment (richard-ramos) [inline libp2p/utils/zeroqueue.nim:48]: Should `Chunk` be a ref object? because otherwise popFirst/addFirst would copy, right?

**Review by richard-ramos — APPROVED** [review]

File: libp2p/utils/zeroqueue.nim (hunk 12 above)
Comment (tersec) [reply libp2p/utils/zeroqueue.nim:33]: As https://status-im.github.io/nim-style-guide/language.result.html notes,
> "expression" returns guarantee that all code branches produce one (and only one) value to be returned

this is something `return foo` doesn't do, so all else equal, expression return is slightly nicer.

File: libp2p/utils/zeroqueue.nim (hunk 16 above)
Comment (tersec) [reply libp2p/utils/zeroqueue.nim:48]: Nim can often use RVO/NRVO in such situations, but it depends.

File: tests/testzeroqueue.nim [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+import unittest2
+import ../libp2p/utils/zeroqueue
+
+proc toSeq(p: pointer, length: int): seq[byte] =
+  let b = cast[ptr UncheckedArray[byte]](p)
+  result = newSeq[byte](length)
```
Comment (tersec) [inline tests/testzeroqueue.nim]: This unnecessarily zero-initializes `result`.

File: libp2p/utils/zeroqueue.nim (hunk 12 above)
Comment (vladopajic) [reply libp2p/utils/zeroqueue.nim:33]: @richard-ramos clear does not return any type?

File: libp2p/utils/zeroqueue.nim (hunk 13 above)
Comment (vladopajic) [reply libp2p/utils/zeroqueue.nim:35]: there is no probably no benefit on inlining this, its not use that often

File: libp2p/utils/zeroqueue.nim (hunk 14 above)
Comment (vladopajic) [reply libp2p/utils/zeroqueue.nim:68]: getting error in that case:
```
Error: type mismatch
Expression: cast[uint](pbytes) + consumed
```

File: libp2p/utils/zeroqueue.nim (hunk 16 above)
Comment (vladopajic) [reply libp2p/utils/zeroqueue.nim:48]: > Should Chunk be a ref object? 

non ref type would copy, but it would be sallow copy, which is negotiable in this case since memory footprint of this type is small (8 bytes). and total number of copies should be vary small, as these operations are not really that much frequent.

however i have made `Chunk` ref object - it goes more in favor for "zero allocation".

File: libp2p/utils/zeroqueue.nim (hunk 12 above)
Comment (richard-ramos) [reply libp2p/utils/zeroqueue.nim:33]: :man_facepalming:  oops
